src/replit_river/v2/session.py

```diff
@@ -0,0 +1,1239 @@
+                    logger.debug(
+                        "_do_ensure_connected: Connection closed during waiting "
+                        "for handshake response",
+                        exc_info=True,
+                    )
+                    raise RiverException(
+                        ERROR_HANDSHAKE,
+                        "Handshake failed, conn closed while waiting for response",
+                    ) from e
+
+                try:
+                    response_msg = parse_transport_msg(data)
+                    if isinstance(response_msg, str):
+                        logger.debug(
+                            "_do_ensure_connected: Ignoring transport message",
+                            exc_info=True,
+                        )
+                        continue
+
+                    break
+                except InvalidMessageException as e:
+                    raise RiverException(
+                        ERROR_HANDSHAKE,
+                        "Got invalid transport message, closing connection",
+                    ) from e
+
+            try:
+                handshake_response = ControlMessageHandshakeResponse(
+                    **response_msg.payload
+                )
+                logger.debug("river client waiting for handshake response")
+            except ValidationError as e:
+                raise RiverException(
+                    ERROR_HANDSHAKE, "Failed to parse handshake response"
+                ) from e
+
+            logger.debug("river client get handshake response : %r", handshake_response)
+            if not handshake_response.status.ok:
+                if handshake_response.status.code == ERROR_CODE_SESSION_STATE_MISMATCH:
+                    do_close()
+
+                raise RiverException(
+                    ERROR_HANDSHAKE,
+                    f"Handshake failed with code {handshake_response.status.code}: {
+                        handshake_response.status.reason
+                    }",
+                )
+
+            # We did it! We're connected!
+            last_error = None
+            rate_limiter.start_restoring_budget(client_id)
+            transition_connected(ws)
+            break
+        except Exception as e:
+            if ws:
+                close_ws_in_background(ws)
+                ws = None
+            last_error = e
+            backoff_time = rate_limiter.get_backoff_ms(client_id)
+            logger.exception(
+                f"Error connecting, retrying with {backoff_time}ms backoff"
+            )
+            await asyncio.sleep(backoff_time / 1000)
+    unbind_connecting_task()
+
+    if last_error is not None:
+        logger.debug("Handshake attempts exhausted, terminating")
+        do_close()
+        raise RiverException(
+            ERROR_HANDSHAKE,
+            f"Failed to create ws after retrying {attempt_count} number of times",
+        ) from last_error
+
+    return None
+
+
+async def _recv_from_ws(
```

Comment:
seems weird that recv_from_ws is responsible for reconnecting

Reply:
it's an [artifact of the `websockets` library](https://websockets.readthedocs.io/en/stable/howto/upgrade.html#open-and-closed), for better or worse. Both `send()` and `recv()` throw `ConnectionClosed`, so it really should be handled in both cases.